plugins/linux-pipewire/portal.h

```diff
@@ -23,7 +23,20 @@
 #include <stdint.h>
 #include <gio/gio.h>
 
-uint32_t portal_get_available_capture_types(void);
+enum portal_capture_type {
```

Comment:
nit, if these are intended to be bitflags they shouldnt be enums. If they are not intended to be bitflags they shouldnt be using different bits.

for example portal_get_available_capture_types now returns an invalid enum member if multiple capture types are available. It should continue to use a uint32_t, and not defining bitsets as enums avoids this kind of problem.

Reply:
This is... tricky. The portal documents these values like this, and they're used both as flags AND enums throughout the stack. In our case, we already have code that uses that in both ways.

Reply:
The portal is defined in dbus syntax, so i dont pretend to know what constitutes a valid enum in that syntax. But in C, we violate the enum definition by setting it to 3, and can see that things like switch statements would fail on such a value. C does not include a type for bitsets so using integers is what we have left.

Reply:
For all intents and purposes, this is an enum. It's an enum that, in one specific situation, acts as a bit flag. This is all expressed in the code: this is declared as an enum (satisfying the enum condition), and as per the portal documentation, each value is defined as a shifted bit (satisfying the bit flag condition).

It must be both, simultaneously. The values must be able to be bit-shifted, otherwise there's no way to express support for multiple capture types through the `AvailableCaptureTypes` D-Bus property. But on every other circumstance, the plugin treats it as an enum, because it never handler nor sends more than one value.

On every situation that this is treated as an enum, receiving 3 must be considered an invalid value.

Reply:
I think you misunderstood me, again the abuse is in using an enum as a bitset (which is common if you treat them like bitflags). And the error is in `portal_get_available_capture_types` which is written to return an enum but can (and will) return 3, which is an invalid value for the enum. Which is then casted away (luckily) to a uint32_t and used correctly by luck.

The only reason to not use bitflag enums is to make such a mistake less likely. Not because bitflag style enums are inherently in violation of any C rules by themselves.

Reply:
Well, okay. I'll be honest with you, in almost 20 years of reading and writing C code, I have never heard of this "bitflags must not be enums" rule. What would you suggest here then? Drop the last commit?

Reply:
1) change the return types for `portal_get_available_capture_types` and `portal_get_available_capture_types` to `uint32_t` to fix the latent bug where we cast invalid values into the enum type.

2) consider if this experience has changed your opinion on using bitflags as enums.

Reply:
Thanks, changed the return types to `uint32_t`

Reply:
> 1. change the return types for `portal_get_available_capture_types` and `portal_get_available_capture_types` to `uint32_t` to fix the latent bug where we cast invalid values into the enum type.

You wrote `portal_get_available_capture_types` twice. Did you mean something else @kkartaltepe ?

Reply:
Sorry i must have fat fingered copy and paste, it was supposed to be `portal_get_available_cursor_modes` but georges got it.

Reply:
Understood.